### DISCUSSION TITLE
How can I get the ID from the current user inside SQL?

### DISCUSSION THREAD

**Discussion by BurSari92** [supabase/supabase#13081] [Questions]
Hi,
I have a simple question, hope someone can help me out.
This is an example function.
This function is getting called by another program (Godot 4.0).
I know there are other ways and even inside godot to get the id, but i need the function to show me the id of the current user.
How can I get the actual id from the user who is calling this function?

create or replace function public.get_text(txt text)
returns text
language plpgsql
security invoker set search_path = public
AS $$
begin
  txt = Concat('Your id is: ',txt);               <---- right here
  return (txt);
end;
$$;

Thanks in advance!

**Comment by GaryAustin1** [comment]
auth.uid() normally shows the signed in user if (jwt sub claim) if you are using Supabase auth.  

You can access any jwt information on database API calls with info from here:  https://postgrest.org/en/stable/api.html#http-context

**Comment by BurSari92** [comment]
Thank you for your reply, but unfortunately it still doesn´t work.
I changed the code into:

**txt = Concat( 'Your id is: ', auth.uid() )**

But i get "null" value in godot instead of the text + id.

The jwt sub claim is a little to difficult for me to understand as a newbie tbh..

**Reply by bombillazo** [reply]
I am having the same issue, I cant get `auth.uid()` to work inside my custom DB function to get the user that has called the function via the supabase client rpc.

**Comment by GaryAustin1** [comment]
@bombillazo 
That highly likely means you don't really have a logged in user, or at least the authorization header being set with a user jwt.
If this is a plpgsql function you can do `raise log 'userid = %', auth.uid();` to check the value in the postgres log.
Also you could do `raise log 'role = %', auth.role();` and if that comes back as "anon" then you do not have a user session set up when you make your rpc call.

**Reply by bombillazo** [reply]
Thanks for knocking some sense to me, I was testing from a SQL query script!!! Not from my front end 🤦 

Now I see the values :D Thanks!

**Reply by whoiscarlo** [reply]
Is it possible though to get `auth.uid()`/ the `current_user` inside postgres? 

I'm trying to make a trigger and `current_user` is returning text instead of a uuid =(

**Reply by GaryAustin1** [reply]
You use auth.uid() in RLS (SQL) or in a DB function as long as the operation was started by rpc initially.

**Reply by whoiscarlo** [reply]
I'm not sure I understand what you mean by 

> in a DB function as long as the operation was started by rpc initially


When I execute the following code in the Supabase SQL Editor: 


```
CREATE OR REPLACE FUNCTION test_changes()
RETURNS TRIGGER AS $$
BEGIN
    INSERT INTO test_table(user_id, changed_at)
    VALUES ( auth.uid(), NOW());
    RETURN NEW;
END;
$$ LANGUAGE plpgsql;

CREATE TRIGGER trigger_test_changes
AFTER UPDATE ON foo_table
FOR EACH ROW EXECUTE FUNCTION test_changes();
```

I receive this error: `null value in column "user_id"`.

But if I understand correctly it's because all of the database updates I am doing are using the [Golang Supabase library](https://github.com/supabase-community/postgrest-go) which does everything through http request? 

If so, do you know of a workaround that would achieve what I'm trying to do?

**Reply by GaryAustin1** [reply]
In the SQL editor you have the option simulate being an authenticated user if use that it will work.
There is no jwt for a request as Postgres user in the SQL and even if there where auth.uid() would be null for that role.
The only time auth.uid() works is if you use a Supabase client and call with a signed in user.  If the user updates your table then your function will pick up the user's id (from the jwt) and auth.uid() would return in it.

If you update the table without a signed in user or directly from the database port with pure SQL then you don't have a signed in user so it is null.

**Reply by whoiscarlo** [reply]
Oh I thought it would work because when a user signs in, I get their **AccessToken** and use that as the apikey when I make http request to make updates to the database. Any updates made, using this **AccessToken** does provide a `auth.uid()` when interacting with any **RLS policies** I've made, so I thought the **Trigger function** would work just the same. 

But I'm assuming it doesn't because **Triggers** are pure SQL so they know anything about `auth.uid()`, while **RLS Policies** have a some kind of  Supabase wrapper around which allows them to know about `auth.uid()`? Or am I completely wrong 😅?

PS: Thank you for taking the time to help me understand this.

**Reply by GaryAustin1** [reply]
You have to use it in the authenticated header as 'bearer jwt' where jwt is the user access token.
Then it will be available for any SQL to use.

**Reply by whoiscarlo** [reply]
Thank you!!! 

I finally found the issue, and per usual it was just user error. I somehow was setting the **Authorization** header with an empty accessToken which was why `auth.iud()` kept turning up as `null`. And the reason why **RLS policies** I had weren't erroring out is because I was using the `service_role` token as my `apikey` in the header, so it was bypassing the policy completed 🤦🏿‍♂️

Which realize that I probably should have never been using the `service_role` key at all? And just stuck with the `anon public`? Is that correct?
